superset-frontend/packages/superset-ui-core/src/color/CategoricalColorScale.ts

```diff
@@ -78,14 +78,18 @@ class CategoricalColorScale extends ExtensibleFunction {
     cleanedValue: string,
   ) {
     // make sure we don't overwrite the origin colors
-    const updatedRange = [...this.originColors];
+    let updatedRange = [...this.originColors];
     // remove the color option from shared color
     sharedColorMap.forEach((value: string, key: string) => {
       if (key !== cleanedValue) {
-        const index = updatedRange.indexOf(value);
-        updatedRange.splice(index, 1);
+        updatedRange = updatedRange.filter(color => color !== value);
       }
     });
+    Object.entries(this.parentForcedColors).forEach(([key, value]) => {
+      if (key !== cleanedValue) {
+        updatedRange = updatedRange.filter(color => color !== value);
+      }
+    }); // remove the color option from forced colors
```

Comment:
The combination of `forEach` and `.filter` operation can be heavy. It should be better using a hashset to filter out the values.

```suggestion
    const updatedRange = new Set(this.originColors);
    // remove the color option from shared color
    sharedColorMap.forEach((value: string, key: string) => {
      if (key !== cleanedValue) {
        updatedRange.delete(value);
      }
    }
    Object.entries(this.parentForcedColors).forEach(([key, value]) => {
      if (key !== cleanedValue) {
        updatedRange.delete(value);
      }
    }
    this.range(updatedRange.size > 0 ? [...updatedRange] : this.originColors);
```

Reply:
oh nice. i will use that